src/exodus/sigmawallet.cpp

```diff
@@ -0,0 +1,597 @@
+    }
+
+    // Verify regenered
+    exodus::SigmaPublicKey pubKey(privKey, DefaultSigmaParams);
+    if (pubKey != pub) {
+        return error("%s: failed to correctly generate mint, pubcoin mismatch", __func__);
+    }
+
+    if (GetSerialId(privKey.serial) != mint.serialId) {
+        return error("%s: failed to correctly generate mint, serial hash mismatch", __func__);
+    }
+
+    return true;
+}
+
+void SigmaWallet::ResetCoinsState()
+{
+    try {
+        CWalletDB walletdb(walletFile);
+
+        ListSigmaMints([&](SigmaMint &m) {
+
+            m.chainState = SigmaMintChainState();
+            m.spendTx = uint256();
+
+            auto priv = GetPrivateKeyFromSeedId(m.seedId);
+            SigmaPublicKey pub(priv, DefaultSigmaParams);
+
+            if (!walletdb.WriteExodusHDMint(
+                SigmaMintId(m.property, m.denomination, pub), m)) {
+
+               throw std::runtime_error("fail to update hdmint");
+            }
+
+        }, false, false);
+
+    } catch (std::runtime_error const &e) {
+        LogPrintf("%s : fail to reset all mints chain state, %s\n", __func__, e.what());
+        throw;
+    }
+}
+
+bool SigmaWallet::SetMintSeedSeen(
+    MintPoolEntry const &mintPoolEntry,
+    uint32_t propertyId,
+    uint8_t denomination,
+    exodus::SigmaMintChainState const &chainState,
+    uint256 const &spendTx)
+{
+    // Regenerate the mint
+    auto const &pubcoin = mintPoolEntry.key;
+    auto const &seedId = mintPoolEntry.seedId;
+    auto seedIndex = GetSeedIndex(seedId);
+
+    SigmaMintId id(propertyId, denomination, pubcoin);
+
+    uint160 serialId;
+
+    // Can regenerate if unlocked (cheaper)
+    if (!pwalletMain->IsLocked()) {
+
+        uint512 seed;
+        GenerateSeed(seedId, seed);
+
+        SigmaPrivateKey coin;
+        if (!SeedToPrivateKey(seed, coin)) {
+            return false;
+        }
+
+        serialId = GetSerialId(coin.serial);
+    } else {
+
+        SigmaMint mint;
+        if (!CWalletDB(walletFile).ReadExodusHDMint(id, mint)) {
+            return false;
+        }
+
+        serialId = mint.serialId;
+    }
+
+    // Create mint object
+    SigmaMint mint(
+        propertyId,
+        denomination,
+        seedId,
+        serialId);
+    mint.chainState = chainState;
+    mint.spendTx = spendTx;
+
+    AddToWallet(mint);
+
+    return true;
+}
+
+SigmaMint SigmaWallet::UpdateMintChainstate(SigmaMintId const &id, SigmaMintChainState const &state)
+{
+    return UpdateMint(id, [&state](SigmaMint &m) {
+        m.chainState = state;
+    });
+}
+
+SigmaMint SigmaWallet::UpdateMintSpendTx(SigmaMintId const &id, uint256 const &tx)
+{
+    return UpdateMint(id, [&tx](SigmaMint &m) {
+        m.spendTx = tx;
+    });
+}
+
+// Mint querying
+bool SigmaWallet::HasMint(SigmaMintId const &id) const
+{
+    CWalletDB walletdb(walletFile);
+    return walletdb.HasExodusHDMint(id);
+}
+
+bool SigmaWallet::HasSerial(secp_primitives::Scalar const &scalar) const
```

Comment:
`scalar` is not indicated what data that need to pass in. Rename it to `serial` will be better.